File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.java
class Solution {
    public boolean isPossible(int[] candies, long k, int val) {
        long cur = 0;
        int i = candies.length - 1;
        while (i >= 0 && cur < k) {
            if (candies[i] >= val) {
                cur += candies[i]/val;
                i--;
            } else {
                return false;
            }
        }
        return cur >= k;
    }

    public int maximumCandies(int[] candies, long k) {
        long total = 0;
        for (int c: candies) total += c;
        if (total < k) return 0;
        Arrays.sort(candies);
        int left = 1, right = (int)(total/k), res = 0;
        while (left <= right) {
            int mid = left + (right-left)/2;
            if (isPossible(candies, k, mid)) {
                res = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return res;
    }
}

File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        def possible(val):
            i = n-1
            cur = 0
            while i >= 0 and cur < k:
                if candies[i] >= val:
                    cur += candies[i]//val
                    i -= 1
                else:
                    return False
            return cur >= k

        n = len(candies)
        Sum = sum(candies)
        if Sum < k:
            return 0
        candies.sort()
        left, right, res = 1, Sum//k, 0
        while left <= right:
            mid = left + (right-left)//2
            if possible(mid):
                left = mid + 1
                res = mid
            else:
                right = mid - 1
        return res